### PR TITLE
Build and test ExecuTorch on PyTorch

### PR DESCRIPTION
**Pull request by huydhn**
This is the first part to start build and test ExecuTorch on PyTorch using a pinned commit.  There will be another PR later to update the pinned commit periodically.

* The pinned commit is in `.ci/docker/ci_commit_pins/executorch.txt` as part of PT Docker image
* I add one simple test `source .ci/scripts/test.sh mv3 cmake xnnpack-quantization-delegation ''`.  More could be added later, in fact, any ET tests on Linux could be run here
* Building and installation vision and audio need to be done in CI after building PyTorch because they will be broken otherwise

Next steps, in sequence:

* [x] Update this pinned commit periodically, similar to https://github.com/pytorch/pytorch/pull/113499
* [ ] Increase ET coverage on PT CI, ideally, we should run all ET pull jobs?
* [x] Switch ExecuTorch's torch, vision, and audio nightly pins to commit pins
* [ ] Update ExecuTorch's torch, vision, and audio commit pins periodically

### CODE REVIEW

File: test/test_model_exports_to_core_aten.py [hunk 1]
```diff
@@ -17,6 +19,7 @@ def _get_ops_list(m: torch.fx.GraphModule):
 
 
 class TestQuantizePT2EModels(TestCase):
+    @pytest.mark.xfail()
```
Comment (huydhn) [inline test/test_model_exports_to_core_aten.py:22]: This test was never run in CI because it lacks the`__main__` part below.  And it's broken atm, and I have created an issue to track that https://github.com/pytorch/pytorch/issues/113506

File: .ci/pytorch/test.sh [hunk 2]
```diff
@@ -991,8 +991,15 @@ test_docs_test() {
 
 test_executorch() {
   # Test torchgen generated code for Executorch.
-  echo "Testing Executorch op registration"
+  echo "Testing ExecuTorch op registration"
```
Comment (huydhn) [inline .ci/pytorch/test.sh:994]: Per the chat with @larryliu0820, I will create an issue to resurrect `test_edge_op_registration` added by https://github.com/pytorch/pytorch/pull/89596 in the new ET world

Comment (huydhn) [reply .ci/pytorch/test.sh:994]: https://github.com/pytorch/pytorch/issues/113507

**Comment by huydhn** [comment]
The failure on the test https://github.com/pytorch/pytorch/actions/runs/6831653384/job/18582195855#step:16:4788 is actually a proof that the test is working as expected.  Atm, there is this change https://github.com/pytorch/pytorch/pull/113115#issuecomment-1806026481 in PyTorch trunk that breaks ExecuTorch.  The diff has been landed internally D51141080 (probably intentionally)

File: .ci/docker/ci_commit_pins/executorch.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+9682172576d5d9a10f3162ad91e0a32b384a3b7c
```
Comment (guangy10) [inline .ci/docker/ci_commit_pins/executorch.txt:1]: When this is a commit of a nightly build, we don't have to build ExecuTorch from source, but would do otherwise correct?

File: .ci/docker/common/install_executorch.sh [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  popd
+}
+
+setup_executorch() {
+  pushd executorch
+  source .ci/scripts/utils.sh
+
+  install_flatc_from_source
+  pip_install .
+  build_executorch_runner "buck2"
```
Comment (guangy10) [inline .ci/docker/common/install_executorch.sh:57]: Would using cmake in release build instead more efficient in CI?

File: .ci/pytorch/build.sh [hunk 5]
```diff
@@ -63,6 +63,12 @@ else
   export LLVM_DIR=/opt/llvm/lib/cmake/llvm
 fi
 
+if [[ "$BUILD_ENVIRONMENT" == *executorch* ]]; then
+  # To build test_edge_op_registration
```
Comment (guangy10) [inline .ci/pytorch/build.sh:67]: What is it?

File: .ci/docker/common/install_executorch.sh [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+install_pip_dependencies() {
+  pushd executorch/.ci/docker
+
+  # Install all Python dependencies
+  pip_install -r requirements-ci.txt
+  # NB: Just need to install the stable version of these domain libraries here
+  # so that we can install executorch.  They will need to be recompiled later
+  # together with torch
+  pip_install torchvision torchaudio
```
Comment (guangy10) [inline .ci/docker/common/install_executorch.sh:46]: What are the failures if we do not install torchvision and torchaudio here? It looks like those stable versions won't be  used and there is no guarantee that they are compatible with the torch.

File: .github/workflows/trunk.yml [hunk 7]
```diff
@@ -174,3 +174,23 @@ jobs:
...
+    uses: ./.github/workflows/_linux-build.yml
+    with:
+      build-environment: linux-jammy-py3-clang12-executorch
+      docker-image-name: pytorch-linux-jammy-py3-clang12-executorch
+      test-matrix: |
+        { include: [
+          { config: "executorch", shard: 1, num_shards: 1, runner: "linux.2xlarge" },
+        ]}
+
+  linux-jammy-py3-clang12-executorch-test:
```
Comment (guangy10) [inline .github/workflows/trunk.yml:189]: Are those merge-blocking jobs? Where is it set?

File: .ci/docker/common/install_executorch.sh (hunk 4 above)
Comment (huydhn) [reply .ci/docker/common/install_executorch.sh:57]: Both will work here as cmake is already included.  I remember ET build still uses buck even when cmake is used (to get the source files), so I think just using buck here is easier

File: .ci/pytorch/test.sh [hunk 8]
```diff
@@ -991,8 +991,25 @@ test_docs_test() {
...
+  # there.  These libraries need to be built here, and not part of the Docker
+  # image because they require the target version of torch to be installed first
+  pip_install --no-use-pep517 --user "git+https://github.com/pytorch/audio.git"
+  pip_install --no-use-pep517 --user "git+https://github.com/pytorch/vision.git"
+
+  echo "Run ExecuTorch regression tests for some models"
+  # NB: This is a sample model, more can be added here
+  export PYTHON_EXECUTABLE=python
+  # shellcheck disable=SC1091
+  source .ci/scripts/test.sh mv3 cmake xnnpack-quantization-delegation ''
```
Comment (guangy10) [inline .ci/pytorch/test.sh:1010]: I'm a bit confused which torch is testing against. IIUC, it should be using the commit of a PR and test if any changes will break ExecuTorch, so the pinned torch is a commit of a PR and torchaudio and torchvision are built on same commit from source. Is it what we doing? I failed to see where the commit of a PR is set and where torch libraries are built

File: .ci/docker/common/install_executorch.sh (hunk 6 above)
Comment (huydhn) [reply .ci/docker/common/install_executorch.sh:46]: Here is the import error https://github.com/pytorch/vision/issues/8101.  Basically, it means that you either use prebuilt nightly binaries for torch, vision, and audio or you need to build all of them from source.  Otherwise, there will be import error when trying to import torchvision.  This means that ET will also need to pin commits not only for torch, but also for vision and audio.  Mixing torch pin commit with nightly vision and audio pre-build binaries won't work

File: .github/workflows/trunk.yml (hunk 7 above)
Comment (huydhn) [reply .github/workflows/trunk.yml:189]: Yup, all the jobs in trunk are merge blocking.  In PyTorch, we are using custom code here https://github.com/pytorch/pytorch/blob/main/.github/scripts/trymerge.py#L1919 to handle the merge, so it's fairly complex.  My plan is to refactor this script to cover ExecuTorch too.

File: .ci/pytorch/test.sh (hunk 8 above)
Comment (huydhn) [reply .ci/pytorch/test.sh:1010]: This one uses:

* PyTorch build from the PR in `linux-jammy-py3-clang12-executorch-build` step.  Basically, the build step builds PyTorch and the build artifacts are copied over to the test job (this script).  So by the time ET test is run, it's guaranteed that PyTorch from the PR is used.  You can see the installation happening here https://github.com/pytorch/pytorch/actions/runs/6837644212/job/18594107600#step:16:132
* It's currently using latest main commit from vision and audio.  The command `pip_install --no-use-pep517 --user "git+https://github.com/pytorch/vision.git` does it.  Once we have pinned commits for vision and audio on Executorch, we can use them instead, thus the TODO part above

File: .ci/docker/common/install_executorch.sh [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+  popd
+}
+
+setup_executorch() {
+  pushd executorch
+  source .ci/scripts/utils.sh
```
Comment (guangy10) [inline .ci/docker/common/install_executorch.sh:53]: This job will rely on some config/scripts in the executorch repo to set up itself. Installing packages and dependencies may be fine but the scripts can be tricky. My concern is that if there are changes in those scripts in ET and it won't be aware of that there is a job in PT repo relies on it, how can we ensure the ET job in PT won't break if someone mess up the scripts/config? It seems like a circular dependency issue

**Review by guangy10 — COMMENTED** [review]
@huydhn Thanks for continue supporting and improving ET stability!

File: .ci/docker/common/install_executorch.sh (hunk 9 above)
Comment (huydhn) [reply .ci/docker/common/install_executorch.sh:53]: True, the reason I use this script is because I don't think it's a good idea to copy/paste the logic from ET CI over here. We have a safe guard though, if someone unknowingly break the script and commit it in ET, next time the pin commit is update, the job here would fail blocking it from landing in PT trunk.  This is the same pattern we are using with XLA here https://github.com/pytorch/pytorch/blob/main/.ci/pytorch/test.sh#L823 where PT CI loads the `common.sh` script from XLA repo.  I guess the solution here is to make sure that `.ci/scripts/utils.sh` stays where it is atm, and backward compatible

Comment (huydhn) [reply .ci/docker/common/install_executorch.sh:53]: So specifically, I need these 2 function from the script `install_flatc_from_source` and `build_executorch_runner "buck2"`.  Do you think it's better to copy them here instead of trying to source the `utils.sh`?

File: .ci/docker/common/install_executorch.sh (hunk 6 above)
Comment (guangy10) [reply .ci/docker/common/install_executorch.sh]: > lol, sorry. I mean the above answer for the question below when I pip install vision and audio in test.sh.
> 
> For this question, may be I don't really need this step. I add them only for `pip_install .` ExecuTorch later to install the python code. Do you know if that step require any version of vision or audio? I think I will get rid of this part to see if CI still works to confirm

Yeah, the error shows that torchvision and torchaudio are needed by torch not ExecuTorch. And it's only needed when build and run something. As far as I know, as deps to ExecuTorch, torchvision and torchaudio are needed by running example models only, so fetching and building torchvision and torchaudio right before running the ET tests would be ideal.

File: .ci/docker/common/install_executorch.sh (hunk 9 above)
Comment (guangy10) [reply .ci/docker/common/install_executorch.sh:47]: I agree with you that current setting is better than copy/paste the logic from ET CI, that would out of maintenance over time. `install_flatc_from_source ` and `build_executorch_runner "buck2"` are relatively stable and testable in ET, I'm less concerning about it. I guess my key questions here are:
  1. Is there any chance the ET job fails in PT due to setup issues (e.g. missing packages, incompatible versions, etc.)?
  2. What is the mitigation process/steps when #1 happens to maximize the availability of the ET job? For example, it would be pretty bad if the fix must be landed to ET first and PT has to wait for a day so that the fix will be picked by the ET nightly build. BC breaking changes could be landed during the downtime.

File: .ci/docker/common/install_executorch.sh (hunk 4 above)
Comment (guangy10) [reply .ci/docker/common/install_executorch.sh]: Yeah, buck is still needed to extract the dep graph. We can only do release build with cmake however.

**Review by guangy10 — COMMENTED** [review]
More follow-up comments

File: .ci/docker/common/install_executorch.sh (hunk 4 above)
Comment (huydhn) [reply .ci/docker/common/install_executorch.sh]: Oh, I see.  I have switched this to cmake

File: .ci/pytorch/test.sh [hunk 10]
```diff
@@ -1000,8 +1000,25 @@ test_docs_test() {
 
 test_executorch() {
   # Test torchgen generated code for Executorch.
-  echo "Testing Executorch op registration"
+  echo "Testing ExecuTorch op registration"
   "$BUILD_BIN_DIR"/test_edge_op_registration
```
Comment (ZainRizvi) [inline .ci/pytorch/test.sh:1004]: Does this line run the actual test? If so, shouldn't this be run after torchvision/audio are installed?

File: test/test_model_exports_to_core_aten.py [hunk 11]
```diff
@@ -1,6 +1,8 @@
 # Owner(s): ["oncall: mobile"]
 import copy
 
+import pytest
```
Comment (ZainRizvi) [inline test/test_model_exports_to_core_aten.py:4]: Is this file relevant to the PR? It doesn't seem like it's one of the executorch tests run

File: .github/workflows/trunk.yml [hunk 12]
```diff
@@ -175,6 +175,26 @@ jobs:
           { config: "force_on_cpu", shard: 1, num_shards: 1, runner: "windows.4xlarge.nonephemeral" },
         ]}
 
+  linux-jammy-py3-clang12-executorch-build:
+    name: linux-jammy-py3-clang12-executorch
```
Comment (ZainRizvi) [inline .github/workflows/trunk.yml:179]: Why did the build step take 1hr 20 mins to run? 
https://github.com/pytorch/pytorch/actions/runs/6868648418/job/18679679827?pr=113364

File: .ci/pytorch/test.sh [hunk 13]
```diff
@@ -1000,8 +1000,25 @@ test_docs_test() {
...
+  # there.  These libraries need to be built here, and not part of the Docker
+  # image because they require the target version of torch to be installed first
+  pip_install --no-use-pep517 --user "git+https://github.com/pytorch/audio.git"
+  pip_install --no-use-pep517 --user "git+https://github.com/pytorch/vision.git"
+
+  echo "Run ExecuTorch regression tests for some models"
+  # NB: This is a sample model, more can be added here
+  export PYTHON_EXECUTABLE=python
+  # shellcheck disable=SC1091
+  source .ci/scripts/test.sh mv3 cmake xnnpack-quantization-delegation ''
```
Comment (ZainRizvi) [inline .ci/pytorch/test.sh:1019]: This generally looks good. Has the executorch team confirmed that the tests being run are a good representative set for what they need?

Given the amount of time spent setting up the environment for this test, the incremental cost of adding a couple more tests would be very cheap

**Review by ZainRizvi — APPROVED** [review]

File: .ci/pytorch/test.sh (hunk 13 above)
Comment (huydhn) [reply .ci/pytorch/test.sh:1019]: > This generally looks good. Has the executorch team confirmed that the tests being run are a good representative set for what they need?

Yeah, I will need to confirm that part with executorch

@guangy10  Do you think this one is ok as the first iteration or we need to test more things here?

File: .github/workflows/trunk.yml (hunk 12 above)
Comment (huydhn) [reply .github/workflows/trunk.yml:179]: Let me take a look.  There are Docker build time which is an one time cost there, but the build does take a bit longer than usual (~55m)

File: test/test_model_exports_to_core_aten.py (hunk 11 above)
Comment (huydhn) [reply test/test_model_exports_to_core_aten.py:4]: The test is supposed to run on Linux in general to provide some cover for Executorch before this change, but it turns out that it wasn't run at all and was broken without anyone knowing.  I could move it to another PR if you think it's unrelated (I agree that it's indeed kind of unrelated)

File: .ci/pytorch/test.sh (hunk 10 above)
Comment (huydhn) [reply .ci/pytorch/test.sh:1004]: Hmm, true.  The test passed, so I guess it doesn't require vision/audio here.  It's better to run it after installing vision and audio anyway, so let me move it down

**Review by malfet — APPROVED** [review]
LGTM, but I wonder if it makes sense to checkout executorch during docker builds, rather than during the test. This way docker image will remain immutable during the pin update (considering that executorch repo is much smaller than pytorch)

File: .github/workflows/trunk.yml (hunk 12 above)
Comment (guangy10) [reply .github/workflows/trunk.yml:179]: > Let me take a look. There are Docker build time which is an one time cost there, but the build does take a bit longer than usual (~55m)

How long is considered normal?

File: .ci/pytorch/test.sh (hunk 13 above)
Comment (guangy10) [reply .ci/pytorch/test.sh:1019]: As we have moved most non-critical and slow-running jobs off pull.yaml in ET, I think we should run most of PR jobs if not all, maybe exclude doc build and linter jobs. I think one benefit is that whenever we add/remove PR jobs in ET, it will automatically reflect to the ET jobs in PT, with reasonable delay (nightly)

File: .github/workflows/trunk.yml (hunk 12 above)
Comment (huydhn) [reply .github/workflows/trunk.yml:179]: With caching, it usually takes half an hour or less for CPU build.  I could simply retry the job first to see if the build time drops

File: .ci/pytorch/test.sh (hunk 13 above)
Comment (huydhn) [reply .ci/pytorch/test.sh:1019]: Ok, that makes sense.  I think I will break that part into a separate PR after this one.  The reason is that running the same number of ET jobs from pull.yml will requires some level of sharding here so that they can be run on more than one runner.  This part is definitely worth another PR in itself.

File: .ci/docker/common/install_executorch.sh (hunk 9 above)
Comment (huydhn) [reply .ci/docker/common/install_executorch.sh:47]: Ah no, PT will checkout and build ET pinned commit directly, there will be no nightly delay.  ET pinned commit here is updated in a controlled way via PR either manual or automated update.  On the automated route, we could have alert if it starts failing, so oncall could address the issue as soon as possible.  On the manual route, whoever update the pinned commit will be blocked if the change on ET break the tests here.

**Review by guangy10 — APPROVED** [review]
Looks good. Exact ET jobs to run will be in a follow-up PR.

File: .github/workflows/trunk.yml (hunk 12 above)
Comment (huydhn) [reply .github/workflows/trunk.yml:179]: Caching is kicking in now, and the build job takes only 16m now https://github.com/pytorch/pytorch/actions/runs/6872890165/job/18692077299
